src/dmd/templateparamsem.d

```diff
@@ -142,25 +142,49 @@ private extern (C++) final class TemplateParameterSemanticVisitor : Visitor
  */
 RootObject aliasParameterSemantic(Loc loc, Scope* sc, RootObject o, TemplateParameters* parameters)
 {
-    if (o)
+    if (!o)
+        return null;
+
+    Expression ea = isExpression(o);
+    RootObject eaCTFE()
+    {
+        sc = sc.startCTFE();
+        ea = ea.expressionSemantic(sc);
+        sc = sc.endCTFE();
+        return ea.ctfeInterpret();
+    }
+    Type ta = isType(o);
+    if (ta && (!parameters || !reliesOnTident(ta, parameters)))
     {
-        Expression ea = isExpression(o);
-        Type ta = isType(o);
-        if (ta && (!parameters || !reliesOnTident(ta, parameters)))
+        Dsymbol s = ta.toDsymbol(sc);
+        if (s)
+            return s;
+        else if (TypeInstance ti = ta.isTypeInstance())
         {
-            Dsymbol s = ta.toDsymbol(sc);
-            if (s)
-                o = s;
+            Type t;
+            const errors = global.errors;
+            ta.resolve(loc, sc, ea, t, s);
+            // if we had an error evaluating the symbol, suppress further errors
+            if (!t && errors != global.errors)
+                return Type.terror;
+            // We might have something that looks like a type
+            // but is actually an expression or a dsymbol
+            // see https://issues.dlang.org/show_bug.cgi?id=16472
+            if (t)
+                return t.typeSemantic(loc, sc);
```

Comment:
This being not covered looks suspicious. same for line 179 and 180.
Does the new code do things that cannot happen ?

Reply:
> This being not covered looks suspicious

Unfortunately the codecov bot is unreliable to the point of uselessness.
This format is copied from `typeSemantic.visitTypeInstance` with the logic changed to fix the issue.

Reply:
> Unfortunately the codecov bot is unreliable to the point of uselessness.

The bot is right, all dmd tests pass when you replace this branch with an `assert(0)`.

Reply:
The bot is still useless because it reports both this and  line 185, one of which must be used for types passed to alias parameters.

Reply:
The test suite passes with the following patch:

```patch
diff --git a/src/dmd/templateparamsem.d b/src/dmd/templateparamsem.d
index 620492fc0..069093188 100644
--- a/src/dmd/templateparamsem.d
+++ b/src/dmd/templateparamsem.d
@@ -171,7 +171,7 @@ RootObject aliasParameterSemantic(Loc loc, Scope* sc, RootObject o, TemplatePara
             // but is actually an expression or a dsymbol
             // see https://issues.dlang.org/show_bug.cgi?id=16472
             if (t)
-                return t.typeSemantic(loc, sc);
+                assert(false, "174");
             else if (ea)
             {
                 return eaCTFE();
@@ -182,7 +182,7 @@ RootObject aliasParameterSemantic(Loc loc, Scope* sc, RootObject o, TemplatePara
                 assert(0);
         }
         else
-            return ta.typeSemantic(loc, sc);
+            assert(false, "185");
     }
     else if (ea)
         return eaCTFE();
```
So the code requiring this change is not included in dmd's test suite.

Reply:
> The bot is still useless because it reports both this and line 185, one of which must be used for types passed to alias parameters.

Maybe add the tests for that then ?